riscv-runtime/src/hash.rs

```diff
@@ -47,40 +47,72 @@ pub fn keccakf(input: &[u64; 25], output: &mut [u64; 25]) {
     }
 }
 
-// Output number of bytes for keccak-256 (32 bytes)
-const W: usize = 32;
-
-/// Keccak function that calls the keccakf machine.
-/// Input is a byte array of arbitrary length and a delimiter byte.
-/// Output is a byte array of length W.
-pub fn keccak(data: &[u8], delim: u8) -> [u8; W] {
-    let mut b = [[0u8; 200]; 2];
-    let [mut b_input, mut b_output] = &mut b;
-    let rate = 200 - (2 * W);
-    let mut pt = 0;
-
-    // update
-    for &byte in data {
-        b_input[pt] ^= byte;
-        pt = (pt + 1) % rate;
-        if pt == 0 {
-            unsafe {
-                keccakf(mem::transmute(&b_input), mem::transmute(&mut b_output));
+pub struct Keccak {
+    state: [u64; 25],
+    buffer: [u8; 136],
+    buffer_pos: usize,
+    rate: usize,
+}
+
+impl Keccak {
+    pub fn v256() -> Self {
+        //RATE = 136
+
+        Self {
+            state: [0u64; 25],
+            buffer: [0u8; 136],
```

Comment:
Being rate a `const`, it could be used here for the array size, instead of repeating the number 126.